### DISCUSSION TITLE
many-to-many broken after 1.4 -> 2.0 upgrade

### DISCUSSION THREAD

**Discussion by jennussbaum** [sqlalchemy/sqlalchemy#10349] [Usage Questions] [answered]
(I've rewritten the original question I started to ask, which I now think was something of an XY question; I believe I'm asking the right thing now.)

I'm more or less new to SQLA. I've inherited a (Flask-based) app that runs fine under 1.4, and am trying to upgrade it to 2.0. I'm getting a number of errors connected to many-to-many relationships, that I cannot seem to solve using the guidelines in the docs.

The app deals with bibliographic records, and there are a number of additional tables and relationships, but they all show similar errors, so I'll just focus on one group. There are three tables, `citation`, `subject`, and `citation_subject`, the latter being an association table that has additional columns, so it needs its own class. The current model setup (with all extraneous columns, methods, etc. stripped out) is:

```
class Citation(db.Model):
    """Citation, many-to-many with Subject and with Author"""
    __tablename__ = 'citation'
    id = db.Column(db.Integer, primary_key=True)
    word = db.Column(db.String(50))
    citationsubject = db.relationship('CitationSubject',
                                      order_by='CitationSubject.created.desc()')
    subject = db.relationship('Subject', secondary="citation_subject",
                              backref=db.backref('citation', lazy='dynamic'))


class Subject(db.Model):
    """Subject, many-to-many with Citation"""
    __tablename__ = 'subject'
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(50))
    citations = db.relationship("CitationSubject", back_populates="subject")


class CitationSubject(db.Model):
    """Linking table for Citation-to-Subject"""
    __tablename__ = 'citation_subject'
    id = db.Column(db.Integer, primary_key=True)
    citation_id = db.Column(db.Integer, db.ForeignKey('citation.id'), default=0, nullable=False)
    subject_id = db.Column(db.Integer, db.ForeignKey('subject.id'), default=0, nullable=False)
    created = db.Column(db.TIMESTAMP, default=db.func.now())
    last_modified = db.Column(db.DateTime, server_default=db.func.now())
    citation = db.relationship('Citation', lazy='joined')
    subject = db.relationship('Subject', backref=db.backref('citation_subject'), lazy='joined', innerjoin=True, order_by='Subject.name')
```
The  most important thing to me is that the app works fine right now (i.e. under 1.4). But the many-to-many setup is very different from what the docs discuss (for example, there's no use of `Mapped` at all, and some of the relationships use the deprecated `backref` while others use `back_references`), and I don't know if that's because things have changed radically under 2.0, or because this was done badly originally and just randomly works anyway, or what. But overall, I'd very much prefer _not_ to make any significant changes unless they're absolutely necessary. I know that the current setup isn't parallel, i.e. there aren't always equivalent relationships declared in every direction (for example, the Citation class has relationships to both CitationSubject and Subject, but the Subject class only has a relationship to CitationSubject (confusingly called "citations"), but not to Citation). But I do know that this version has the relationships the app needs, with things named as the app uses the names.

When I start the app, I get a number of errors of this sort:

```
relationship 'CitationSubject.subject' will copy column subject.id to column citation_subject.subject_id, which conflicts with relationship(s): 'Citation.subject' (copies subject.id to citation_subject.subject_id), 'Subject.citation' (copies subject.id to citation_subject.subject_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="citationsubject"' to the 'Subject.citation' relationship. (Background on this warning at: https://sqlalche.me/e/20/qzyx) (This warning originated from the `configure_mappers()` process, which was invoked automatically in response to a user-initiated operation.)
```
There are several blocks of such errors, with different relationships highlighted; I can provide others if necessary. In some cases, I don't actually know where the relationships come from; one error complains about "relationship 'Subject.citation'", but the relationship in my `Subject` class is to `citations`, with the -s; another error shows 'Subject.citation_subject' which I likewise don't declare anywhere.

I have read the discussion of this specific error at https://docs.sqlalchemy.org/en/20/errors.html#error-qzyx, as linked, and I tried to add `back_populates` in various locations as it suggests, but most of the errors still remain. The original question I had started to ask was solely about this, but I know think that the problem is more with my overall many-to-many setup itself, rather than a need to add one or two specific `back_populates`  parameters. 

Again, I'd prefer not to make radical changes, but I am willing to do what's necessary to get things right. Thank you for your help.

**Comment by zzzeek** [comment, marked as answer]
hey there -

that message is not an error message, it's a warning.  the program will still continue to run despite the warning, which is merely pointing out something that was already true of the application which states that there are more than one `relatoinship()` which could be manipulated into writing data into a particular column.

the warning itself includes the exact steps you need to make the warning go away and have your program continue to run in exactly the same way it does in 1.4:

```
To silence this warning, add the parameter 'overlaps="citationsubject"' to the 'Subject.citation' relationship.
```


that means this:

```py
class CitationSubject(db.Model):
    """Linking table for Citation-to-Subject"""
     # ...

    subject = db.relationship('Subject', backref=db.backref('citation_subject'), lazy='joined', innerjoin=True, overlaps="citationsubject", order_by='Subject.name')

```

since you are upgrading a production application from 1.4 to 2.0, you can add that directive and at that point you can be done.  

The pattern that you have in use however is the one that this warning was designed to accommodate; you have the association object pattern in use, while **at the same time** the same mapped table is referenced as ``secondary`` in another relationship.   The warning rightly points out that if you were to create `CitationSubject` objects while **at the same time** appended new records to `Citation.subject`, it would create duplicate rows in the database.      So if you wanted to be more finely tuned, you could place viewonly=True in those relationships like this:

```py
class Citation(db.Model):
    """Citation, many-to-many with Subject and with Author"""
    # ...

    subject = db.relationship('Subject', secondary="citation_subject", viewonly=True,
                              backref=db.backref('citation', lazy='dynamic'))
```

that would also make the warning go away.

It also seems like `CitationSubject.citation` and `Citation.citationsubject` need to be assocaited together with back_populates, that might also be producing a warning otherwise:

```py
class Citation(db.Model):
    """Citation, many-to-many with Subject and with Author"""
    # ...

    citationsubject = db.relationship('CitationSubject',
                                      back_populates='citation',
                                      order_by='CitationSubject.created.desc()')



class CitationSubject(db.Model):
    """Linking table for Citation-to-Subject"""
    # ...


    citation = db.relationship('Citation', lazy='joined', back_populates='citationsubject')

```

but by all means, just getting the app to work, read each warning and apply the "overlaps" directive as written.  then you are back to normal



>  But the many-to-many setup is very different from what the docs discuss (for example, there's no use of Mapped at all, and some of the relationships use the deprecated backref while others use back_references), and I don't know if that's because things have changed radically under 2.0, or because this was done badly originally and just randomly works anyway, or what. 

all of that is fine here, it's using standard 1.x patterns which are fully supported under 2.0.

**Reply by jennussbaum** [reply]
Wow, thank you so much for this fast and detailed answer.

It looks like my original question may indeed have been the right one! I assumed that everything was broken. Your suggestions all work perfectly. I confess I wasn't sure when the "overlaps" or "viewonly" directives were required so I was afraid to try them.... Now on to the additional tables/classes that I didn't mention originally!

I'll put aside the question of how to rewrite this to fully use 2.0 patterns to another day.

**Comment by CaselIT** [comment]
> There are several blocks of such errors, with different relationships highlighted; I can provide others if necessary. In some cases, I don't actually know where the relationships come from; one error complains about "relationship 'Subject.citation'", but the relationship in my Subject class is to citations, with the -s; another error shows 'Subject.citation_subject' which I likewise don't declare anywhere.

You *are* declaring them though: 
- `backref=db.backref('citation', lazy='dynamic'` this creates a *new* relationship in `Subject` called `citation` with `lazy='dynamic'` as parameter
- `backref=db.backref('citation_subject')`: this creates a *new* relationship in `Subject` called `citation_subject`


this is the main reason why backref is now legacy, since it creates things that are hard to find

Since those relationships have duplicated version that are explicitly declared, it probably makes sense consolidate the code to use only one copy if possible, or apply `viewonly=True` to the ones that are used only as read only, like maybe the one with `lazy='dynamic'`
